docs/client-specifications/error-handling.md

```diff
@@ -89,13 +95,84 @@ interface MomentoGrpcErrorDetails {
+
+## Languages that support pattern matching
+
+Examples in this section are based on C#.  Kotlin, Rust, and possibly other languages will have better support for
+pattern matching.
+
+### `CreateCache` and similar operations
+
+The `CreateCache` return value should be of type `CreateCacheResponse`, and should have three sub-types: `Success`,
+`AlreadyExists`, and `Error`.  Pattern matching is used to access the sub-types and their corresponding properties.
+
+```csharp
+var createCacheResponse = await client.CreateCacheAsync("myCache");
+if (createCacheResponse is CreateCacheResponse.Success createCacheSuccess)
+{
+  Console.WriteLine("Cache created");
+}
+else if (createCacheResponse is CreateCacheResponse.AlreadyExists createCacheAlreadyExists)
+{
+  Console.WriteLine("Already exists");
+}
+else if (createCacheResponse is CreateCacheResponse.Error createCacheError) {
+  Console.WriteLine($"Error {createCacheError.ErrorCode} creating cache: {createCacheError}");  
+} else {
+  Console.WriteLine($"Something unexpected happened while attempting to create cache: ${createCacheResponse}");
+}
+```
+
+### `Get` and similar operations
+
+The `Get` return value should be of type `CacheGetResponse`, and should have three sub-types: `Hit`,
+`Miss`, and `Error`.  Pattern matching is used to access the sub-types and their corresponding properties.
+
+```csharp
+var getResponse = await client.GetAsync("myCache", "myKey");
+if (getResponse is CacheGetResponse.Hit hitResponse)
+{
+  Console.WriteLine("Cache hit; value: ${getResponse.ValueString}");
+}
+else if (getResponse is CacheGetResponse.Miss missResponse)
+{
+  Console.WriteLine("Cache miss!");
+}
+else if (getResponse is CacheGetResponse.Error errorResponse)) {
+  Console.WriteLine($"Error {errorResponse.ErrorCode}: {errorResponse}");
+} else {
+  Console.WriteLine($"`Something unexpected occurred while attempting to get key 'myKey' from cache 'myCache': {getResponse}");
+}
+```
+
+### `Set` and similar operations
+
+The `Set` return value should be of type `CacheSetResponse`, and should have two sub-types: `Success`,
+and `Error`.  Pattern matching is used to access the sub-types and their corresponding properties.
+
+```csharp
+var setResponse = await client.SetAsync("myCache", "myKey", "myValue");
+if (setResponse is CacheSetResponse.Hit successResponse)
+{
+  Console.WriteLine("Cache set successful");
+}
+else if (setResponse is CacheSetResponse.Error errorResponse)) {
+  Console.WriteLine($"Error {errorResponse.ErrorCode}: {errorResponse}");
+} else {
+  Console.WriteLine($"`Something unexpected occurred while attempting to set key 'myKey' in cache 'myCache': {setResponse}");
+}
+```
+
+## Languages that do not support pattern matching
+
+In languages that don't support pattern matching (PHP, TypeScript, etc.) we will use a status enum or `as` methods
+to cast to the appropriate response type.
```

Comment:
Is it worth adding an `as` method example here? It's different enough from this enum-based approach to make it potentially confusing for someone following along IMO.

Reply:
yeah, that's a good callout.  tbh I was kind of waiting to see what we end up doing for the javascript one since we haven't updated it yet.  Once someone starts messing with it I don't know if we are going to end up going with `as` there or have the nullable fields etc.  So I was thinking of updating this after that point in time.  wdyt?

Reply:
I'm again considering adding `success?`, `error?`, `already_exists?` methods in Ruby to make calling more convenient.

```
case response = client.create_cache("test_cache")
when Momento::Response::CreateCache::Success
  # huzzah
when Momento::Response::CreateCache::Error::AlreadyExists
  # ignore if it already exists
when Momento::Response::CreateCache::Error
  # uh oh
end
```

```
case response = client.create_cache("test_cache")
when response.success?
  # huzzah
when response.already_exists?
  # ignore if it already exists
when response.error?
  # uh oh
end
```

Reply:
i like that!